python/pylibcudf/tests/test_lists.py

```diff
@@ -161,28 +160,59 @@ def test_reverse(list_column):
     assert_column_eq(expect, got)
 
 
-# https://github.com/rapidsai/cudf/issues/19346
-@pytest.mark.flaky(reruns=5)
-def test_segmented_gather(test_data):
-    list_column1, list_column2 = test_data[0]
-
-    plc_column1 = plc.Column.from_arrow(pa.array(list_column1))
-    plc_column2 = plc.Column.from_arrow(pa.array(list_column2))
-
-    got = plc.lists.segmented_gather(plc_column2, plc_column1)
-
-    expect = pa.array([[8, 9], [14], [0], [0, 0]])
-
-    assert_column_eq(expect, got)
-
-
-def test_extract_list_element_scalar(list_column):
-    plc_column = plc.Column.from_arrow(pa.array(list_column))
-
-    got = plc.lists.extract_list_element(plc_column, 0)
-    expect = pc.list_element(list_column, 0)
+@pytest.mark.parametrize(
+    "bounds_policy",
+    [
+        plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+        plc.copying.OutOfBoundsPolicy.NULLIFY,
+    ],
+    ids=["DONT_CHECK", "NULLIFY"],
+)
+def test_segmented_gather(
+    bounds_policy: plc.copying.OutOfBoundsPolicy,
+) -> None:
+    # Cases to check, which map to rows in the Column:
+    # 1. All in-bounds
+    # 2. Empy source, non-empty indexer (all out of bounds)
+    # 3. Empty source, empty indexer (all in bounds)
+    # 4. Non-empty source, empty indexer (all in bounds)
+    # 5. Nullable source, non-empty indexer (mix of out-of-bounds and in-bounds)
+
+    source_column = plc.Column.from_arrow(
+        pa.array([[10, 20], [], [], [30], [None, 50, 30, 20]])
+    )
+    gather_map_list = plc.Column.from_arrow(
+        pa.array([[-2, -1, 0, 1], [0, -1], [], [], [0, 1, 3, -1, 9]])
+    )
+    got = plc.lists.segmented_gather(
+        source_column, gather_map_list, bounds_policy
+    )
 
-    assert_column_eq(expect, got)
+    if bounds_policy == plc.copying.OutOfBoundsPolicy.DONT_CHECK:
+        # using -1 as a placeholder for out-of-bounds values
+        expect = pa.array(
+            [[10, 20, 10, 20], [-1, -1], [], [], [None, 50, 20, 20, -1]]
+        )
+        # We can't check the exact values because the behavior is undefined for out-of-bounds indices
+        # when bounds_policy is DONT_CHECK.
+        result = plc.interop.to_arrow(got)
+        assert len(result) == 5
+        assert len(result[0]) == 4
+        assert len(result[1]) == 2
+        assert len(result[2]) == 0
+
+        # these are all in-bounds and we can check exactly
+        assert expect[:1].equals(result[:1])
+        assert expect[2:-1].equals(result[2:-1])
+
+        # these include OOB indices
+        assert len(result[1]) == 2
+        assert len(result[4]) == 5
```

Comment:
Sorry but I don't like this as much. I'd rather have two tests (or one test with different gather indices) so we can just call `assert_column_eq`. And we don't need to check `DONT_CHECK` for OOB indices since I think that is tested in the C++ layer.

Reply:
Plus the bounds policy is documented now in pylibcudf. I think we'd have root-caused the failing test earlier if it had been.